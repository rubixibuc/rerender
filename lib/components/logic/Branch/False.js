import React from "react";
import PropTypes from "prop-types";

const render = ({ children }) => <React.Fragment>{children}</React.Fragment>;

render.propTypes = {};

export default render;
