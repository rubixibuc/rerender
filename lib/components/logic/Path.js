import React from "react";
import PropTypes from "prop-types";

const render = ({ children, condition }) => (
  <React.Fragment>{children}</React.Fragment>
);

render.propTypes = {
  children: PropTypes.node,
  condition: PropTypes.oneOfType([PropTypes.bool, PropTypes.func])
};

render.displayName = "Rerender(Path)";

export default render;
