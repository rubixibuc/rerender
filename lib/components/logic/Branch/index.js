import React from "react";
import PropTypes from "prop-types";

import Both from "./Both";
import False from "./False";
import True from "./True";

const render = ({ children, condition }) =>
  children
    ? React.Children.map(
        children,
        child =>
          (condition() && child.type === True) ||
          (!condition() && child.type === False) ||
          child.type === Both
            ? child
            : null
      )
    : null;

const TrueFalsePropType = (
  props,
  propName,
  componentName,
  location,
  propFullName
) =>
  props[propName].type !== True &&
  props[propName].type !== False &&
  props[propName] !== Both
    ? new Error(
        "Invalid prop `" +
          propFullName +
          "` supplied to" +
          " `" +
          componentName +
          "`. Validation failed."
      )
    : undefined;

render.propTypes = {
  children: PropTypes.oneOfType([
    TrueFalsePropType,
    PropTypes.arrayOf(TrueFalsePropType)
  ]),
  condition: PropTypes.func.isRequired
};

export default render;
