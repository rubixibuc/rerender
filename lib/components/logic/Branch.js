import React from "react";
import PropTypes from "prop-types";

import Path from "./Path";

const render = ({ children, first }) => {
  let met = false;

  return children
    ? React.Children.map(children, child => {
        if (
          !met &&
          (!("condition" in child.props) ||
            (typeof child.props.condition === "function"
              ? child.props.condition()
              : child.props.condition))
        ) {
          met = first && true;
          return child;
        }
        return null;
      })
    : null;
};

const PathPropType = (props, propName, componentName, location, propFullName) =>
  props[propName].type !== Path
    ? new Error(
        "Invalid prop `" +
          propFullName +
          "` supplied to" +
          " `" +
          componentName +
          "`. Validation failed."
      )
    : undefined;

render.propTypes = {
  children: PropTypes.oneOfType([
    PathPropType,
    PropTypes.arrayOf(PathPropType)
  ]),
  first: PropTypes.bool
};

render.defaultProps = {
  first: true
};

render.displayName = "Rerender(Branch)";

export default render;
