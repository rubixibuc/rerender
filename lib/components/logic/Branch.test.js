import React from "react";
import renderer from "react-test-renderer";

import Branch from "./Branch";
import Path from "./Path";

test("should render nothing when children are undefined", () => {
  const component = renderer.create(<Branch />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render nothing when children are undefined and taking every path", () => {
  const component = renderer.create(<Branch first={false} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render first Path (2) with true condition in Branch", () => {
  const component = renderer.create(
    <Branch>
      <Path condition={false}>1</Path>
      <Path condition={true}>2</Path>
      <Path condition={true}>3</Path>
      <Path condition={false}>4</Path>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render first Path (2) with function condition returning true in Branch", () => {
  const component = renderer.create(
    <Branch>
      <Path condition={false}>1</Path>
      <Path condition={() => true}>2</Path>
      <Path condition={true}>3</Path>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render all Paths (2, 3) with true condition in Branch", () => {
  const component = renderer.create(
    <Branch first={false}>
      <Path condition={false}>1</Path>
      <Path condition={true}>2</Path>
      <Path condition={true}>3</Path>
      <Path condition={false}>4</Path>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render no Paths because none match", () => {
  const component = renderer.create(
    <Branch first={false}>
      <Path condition={false}>1</Path>
      <Path condition={false}>2</Path>
      <Path condition={false}>3</Path>
      <Path condition={false}>4</Path>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render no Paths because none match when taking every path", () => {
  const component = renderer.create(
    <Branch first={false}>
      <Path condition={false}>1</Path>
      <Path condition={false}>2</Path>
      <Path condition={false}>3</Path>
      <Path condition={false}>4</Path>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render Path (3) with no condition because all other Path conditions are false", () => {
  const component = renderer.create(
    <Branch first={false}>
      <Path condition={false}>1</Path>
      <Path condition={false}>2</Path>
      <Path>3</Path>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
