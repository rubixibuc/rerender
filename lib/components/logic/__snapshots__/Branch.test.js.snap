// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render all Paths (2, 3) with true condition in Branch 1`] = `
Array [
  "2",
  "3",
]
`;

exports[`should render first Path (2) with function condition returning true in Branch 1`] = `"2"`;

exports[`should render first Path (2) with true condition in Branch 1`] = `"2"`;

exports[`should render no Paths because none match 1`] = `null`;

exports[`should render no Paths because none match when taking every path 1`] = `null`;

exports[`should render nothing when children are undefined 1`] = `null`;

exports[`should render nothing when children are undefined and taking every path 1`] = `null`;

exports[`should render path with no condition because all other Path conditions are false 1`] = `"3"`;
